### ISSUE TITLE
Users are deactivated but can log in when ID Attribute changes

### ISSUE THREAD

**Issue by MrProteus** [mattermost/mattermost#25816]
#### Summary
When the capitalization of the ID attribute value changes, the Mattermost account is deactivated on every LDAP sync and the session expired, but the user can still log in. 
See also: <https://forum.mattermost.com/t/ldap-sync-deactivates-user-unexpectedly/5496>
<https://mattermost.atlassian.net/browse/MM-11525>

#### Steps to reproduce
Ubuntu 22.04
Mattermost 9.2.3
LDAP Sync from Active Directory with sAMAccountName as ID attribute and login ID attribute
1. sync existing AD user into Mattermost
2. log in as user into Mattermost
3. change the capitalization of sAMAccountName, i.e. Auser to auser
4. sync LDAP in Mattermost
5. user is deactivated in Mattermost and their session expired
6. log in with ANY capitalization of sAMAccountName as login ID attribute
7. user is successfully logged into their old account (i.e. retains their conversations and channel memberships)
8. sync LDAP again
9. user is again deactivated and their session expired

#### Expected behavior
EITHER
change of ID attribute causes deactivation of account and user cannot log in anymore
OR
user can still log in and account is not deactivated (and their sessions don't expire)

#### Observed behavior (that appears unintentional) 
LDAP syncs after change of ID attribute value deactivate the account and expire all sessions; however, the user can still log in with their credentials and retain the same Mattermost account (i.e. keeps their conversations and channel memberships). They stay logged in until another LDAP sync is performed, upon which the log shows that the user is deactivated again (and logged out of any sessions). This matches the bevahior described in this forum post: <https://forum.mattermost.com/t/ldap-sync-deactivates-user-unexpectedly/5496>, however it is not clear to me whether the user in that case was still able to log in. 
It seems unintentional that a user that is marked as deactivated can log in. While it is stated in <https://mattermost.atlassian.net/browse/MM-11525> that the ID attribute being case sensitive is intended, this behavior makes it hard to debug user reports like "I am logged out of my account every day". Alternatively, if a change in the capitalization of the ID attribute should still result in an user being able to log into the same Mattermost account, the account should not be deactivated and their sessions should not expire. 

#### Possible fixes
As a manual workaround, if the AuthData column value in the user database is changed to match the ID attribute including case, the Mattermost account is not deactivated on LDAP syncs anymore.

**Comment by amyblais** [comment]
Opened https://mattermost.atlassian.net/browse/MM-56392.

**Comment by hanzei** [comment]
Thanks for the report @MrProteus.

I can reproduce that changing the `sAMAccountName` results in the old user getting deactivated. It's not clear to me if that is the expected behaviour. I need to confirm that with the Security team.

But I cannot reproduce that the user can log in using ANY capitalization of sAMAccountName. In my testing, after the `sAMAccountName`  was changed, the user could not log in using _any_ capitalization of sAMAccountName.

What Mattermost version are you using?

**Comment by MrProteus** [comment]
We are using version 9.2.3, Database Schema Version: 113, Build Number: 6961628977. 
I just double-checked - the user can log in with different capitalizations even when `sAMAccountName `and `ID attribute value` match exactly (i.e. `sAMAccountName` and `ID attribute value` are `auser` but the user can log in using `Auser`, `AUSER` or `aUsEr`).

**Comment by MrProteus** [comment]
Yes, the user can log in with different capitalizations both before and after changing `sAMAccountName` in AD. 
`IdAttribute` and `LoginIdAttribute` are both set to `sAMAccountName`.

**Comment by MrProteus** [comment]
Yes.

**Comment by hanzei** [comment]
I'm still unable to reproduce the issue. Steps 1 to 5 work as described, but on Step 6 the login (with any capitalization) fails with an error. The error is thrown because Mattermost tries to create a new user account (because `sAMAccountName` does not match with the existing account), which fails because usernames are case insensitive in MM and hence, an account with the same username already exists (which is the "old", deactivated account)

To debug your issue further, could you please access your DB directly and get the `authdata` field from the deactivated account? Please run `SELECT authdata from users WHERE id = 'XYZ'` while replacing `XYZ` with the userID of the account.

**Comment by MrProteus** [comment]
I started out with `sAMAccountName` in AD and `AuthData` being `auser` -- everything works as expected. I then changed `sAMAccountName` to `Auser` in AD and synced; inspecting the DB shows that `AuthData` and `Username` stay `auser`, `DeleteAt` now contains a non-zero value (that matches the value of `UpdateAt`).  

Logging in with `auser` (or other spellings like `AUSER`) succeeds, `AuthData` and `Username` are still `auser`; `DeleteAt` is again 0 and `UpdateAt` is increased. 

Just to rule this out as an error source, the `Id` column value in the DB is a random string starting with `etd73` - it is not related to `sAMAccountName` (but `IdAttribute` is set to `sAMAccountName` in config.json).

**Comment by MrProteus** [comment]
We have since migrated from MySQL to Postgres and upgraded Mattermost several times and cannot reproduce the error anymore. It's hard to confirm this now, but my suspicion is that a column in one of the tables in MySQL was set to case-insensitive - that would explain why database calls would still return matches, but string-comparisons in the Mattermost would not. 

Closing this issue as it can no longer be reproduced.
